### ISSUE TITLE
🐛 [firestore_odm] Not generating collection references

### ISSUE THREAD

**Issue by cedvdb** [firebase/flutterfire#10956]
## Bug report

Seems like odm generator does not generate collection references. I'm not the only one stumbling upon this so it's probably not an issue with my setup.

### Steps to reproduce

Repro: https://github.com/cedvdb/test_firestore_odm

run:`flutter pub run build_runner build — delete-conflicting-outputs`

nothing is generated.

--------------------------------

Edit: seems like the file is generated in .dart-tools/

**Comment by gmarizy** [comment]
@rrousselGit stated it's related to https://github.com/dart-lang/graphs/issues/86. 

Indeed, the workaround found in this issue does work: 
```
dependency_overrides:
  graphs: 2.2.0
```

**Comment by rrousselGit** [comment]
Indeed. Closing as it's not really an ODM bug

**Comment by gmarizy** [comment]
@rrousselGit it may be better to keep this issue open until the problem is solved on graph side. While there is nothing wrong with ODM, every user who use flutter pub upgrade is going to run into this problem. Since not everyone is going to search in closed issue, you will soon have a duplicate issue on the same topic. Open issues have a better visibility.

**Comment by borgoat** [comment]
I've been facing the same issue, but even downgrading `graphs` as mentioned in the comment above[^1] didn't work for me... 

I believe the override was applied correctly since a `pub get` shows:

```
! graphs 2.2.0 (overridden) (2.3.0 available)
```

Initially, `models.firebase.g.part` and `models.json_serializable.g.part` are generated, but then they are not correctly merged to form `models.g.dart`[^2]

I attempted both with Dart 3 and the previous version, but nothing changed.


[^1]: https://github.com/firebase/flutterfire/issues/10956#issuecomment-1545862884
[^2]: https://gist.github.com/borgoat/e96330d4ef65be89d542456ed48706f0

**Comment by cedvdb** [comment]
@borgoat run flutter clean before flutter pub get it solved it for me

**Comment by gmarizy** [comment]
Problem came back despite overriding graphs dependency. After investigation, it was related to updating build_runner version from 2.4.2 to 2.4.4. Indeed, 2.4.4 changelog is "Use a stable order for builders without an order defined by dependencies.", which is closely related to the linked graphs issue. 

My working configuration is now:
```
dependency_overrides:
  graphs: 2.2.0

dev_dependencies:
  flutter_test:
    sdk: flutter
  build_runner: 2.4.3
  ...
```

**Comment by tristanjclarke** [comment]
After upgrading to Flutter version `3.10.0` (the previous Flutter version was `3.7.8`), and running the below commands the `CollectionReference` is not being generated.

- `flutter clean`
- `flutter pub get`
- `flutter pub run build_runner build --delete-conflicting-outputs`

I'm running the following config in `pubspec.yaml`:

```
dependencies:
  cloud_firestore: ^4.7.0
  cloud_firestore_odm: ^1.0.0-dev.58

dev_dependencies:
  build_runner: ^2.4.3
  cloud_firestore_odm_generator: ^1.0.0-dev.58
```

Example code to be generated is:

```dart
@Collection<FirestoreClass>('class_listing')
@freezed
class FirestoreClass with _$FirestoreClass {
  static const firstArgument = 1;
  static const secondArgument = 2;

  const FirestoreClass._();

  static const defaultValue = FirestoreClass();

  @FirestoreSerializable(fieldRename: FieldRename.kebab)
  @HiveType(typeId: HiveTypes.class)
  const factory FirestoreClass({
    @HiveField(FirestoreClass.fieldFirstArgument) List<String>? firstArgument,
    @HiveField(FirestoreClass.fieldSecondArgument) List<String>? secondArgument,
  }) = _FirestoreClass;

  factory FirestoreClass.fromJson(Map<String, dynamic> json) => _$$_FirestoreClassFromJson(json);
}

final firestoreClassCollectionReference = FirestoreClassCollectionReference();
```

I've tried downgrading back to the previous Flutter version and dependencies but still no luck in generating code now...

**Comment by eddiebeazer** [comment]
> After upgrading to Flutter version `3.10.0` (the previous Flutter version was `3.7.8`), and running the below commands the `CollectionReference` is not being generated.
> 
> 
> 
> - `flutter clean`
> 
> - `flutter pub get`
> 
> - `flutter pub run build_runner build --delete-conflicting-outputs`
> 
> 
> 
> I'm running the following config in `pubspec.yaml`:
> 
> 
> 
> ```
> 
> dependencies:
> 
>   cloud_firestore: ^4.7.0
> 
>   cloud_firestore_odm: ^1.0.0-dev.58
> 
> 
> 
> dev_dependencies:
> 
>   build_runner: ^2.4.3
> 
>   cloud_firestore_odm_generator: ^1.0.0-dev.58
> 
> ```
> 
> 
> 
> Example code to be generated is:
> 
> 
> 
> ```dart
> 
> @Collection<FirestoreClass>('class_listing')
> 
> @freezed
> 
> class FirestoreClass with _$FirestoreClass {
> 
>   static const firstArgument = 1;
> 
>   static const secondArgument = 2;
> 
> 
> 
>   const FirestoreClass._();
> 
> 
> 
>   static const defaultValue = FirestoreClass();
> 
> 
> 
>   @FirestoreSerializable(fieldRename: FieldRename.kebab)
> 
>   @HiveType(typeId: HiveTypes.class)
> 
>   const factory FirestoreClass({
> 
>     @HiveField(FirestoreClass.fieldFirstArgument) List<String>? firstArgument,
> 
>     @HiveField(FirestoreClass.fieldSecondArgument) List<String>? secondArgument,
> 
>   }) = _FirestoreClass;
> 
> 
> 
>   factory FirestoreClass.fromJson(Map<String, dynamic> json) => _$$_FirestoreClassFromJson(json);
> 
> }
> 
> 
> 
> final firestoreClassCollectionReference = FirestoreClassCollectionReference();
> 
> ```
> 
> 
> 
> I've tried downgrading back to the previous Flutter version and dependencies but still no luck in generating code now...
> 
> 
> 
> FYI - I'm not using the `graphs` dependency.


I was not using graphs as a direct dependency but the override suggestion from the first comment worked for me while using the most up to date dependencies at the time. I'd suggest giving it a try

**Comment by gmarizy** [comment]
@tristanjclarke you don't use `graphs` directly, but cloud_firestore_odm does. Also, `build_runner: ^2.4.3` mean that you instruct to use latest update of the lib except major one (a major one is one that update the first digit). So ^2.4.3 resolve to ^2.4.4, which mess with ODM too. Use the configuration given above and you should be fine.

**Comment by tristanjclarke** [comment]
Thanks for the support @gmarizy and @eddiebeazer. This has resolved the problem for me in the mean time 🙏

**Comment by gmarizy** [comment]
Starting with version 1.0.0-dev.59, we no longer need to override graphs dependency to 2.2.0. ODM is still not compatible with graphs 2.3.0, but cloud_firestore_odm_generator now define in it's build.yaml : `graphs: '>=2.0.0 <2.3.0'`.

The problem with build_runner 2.4.4 remain, meaning we should still fix it's version: `build_runner: 2.4.3`.

**Comment by rrousselGit** [comment]
I'm back from holidays. Will fix today.

**Comment by cedvdb** [comment]
Note that I don't think build_runner 2.4.4 is compatible still, unless that was a quircks in my run.

as @gmarizy mentioned it has to be fixed to 2.4.3
